src/backend/env/mod.rs

```diff
@@ -989,26 +1004,79 @@ impl State {
+    pub fn create_invite(
+        &mut self,
+        principal: Principal,
+        credits: Credits,
+        credits_per_user_opt: Option<Credits>,
+        realm_id: Option<RealmId>,
+    ) -> Result<(), String> {
+        let credits_per_user = credits_per_user_opt.unwrap_or(credits);
+        if credits % credits_per_user != 0 {
+            return Err("Credits per user are not a multiplier of credits".into());
+        }
         let min_credits = CONFIG.min_credits_for_inviting;
-        let user = self
-            .principal_to_user_mut(principal)
-            .ok_or("no user found")?;
-        if credits < min_credits {
+        let user = self.principal_to_user(principal).ok_or("no user found")?;
+        if credits_per_user < min_credits {
             return Err(format!(
                 "smallest invite must contain {} credits",
                 min_credits
             ));
         }
-        if user.credits() < credits {
-            return Err("not enough credits".into());
-        }
+
+        invite::validate_realm_id(self, realm_id.as_ref())?;
+        invite::validate_user_invites_credits(self, user, credits, None)?;
+
         let mut hasher = Sha256::new();
         hasher.update(principal.as_slice());
         hasher.update(time().to_be_bytes());
         let code = format!("{:x}", hasher.finalize())[..10].to_string();
         let user_id = user.id;
-        self.invites.insert(code, (user_id, credits));
+        let invite = Invite::new(credits, credits_per_user, realm_id, user_id);
+        self.invite_codes.insert(code, invite);
+
+        Ok(())
+    }
+
+    pub fn update_invite(
+        &mut self,
+        user_id: UserId,
+        invite_code: String,
+        credits: Option<Credits>,
+        realm_id: Option<RealmId>,
+    ) -> Result<(), String> {
+        if credits.is_none() && realm_id.is_none() {
+            return Err("Update is empty".into());
+        }
+        let Some(user) = self.users.get(&user_id) else {
+            return Err("User not found".into());
+        };
+
+        invite::validate_realm_id(self, realm_id.as_ref())?;
+
+        let Invite {
+            credits: invite_credits,
+            credits_per_user,
+            ..
+        } = self
+            .invite_codes
+            .get(&invite_code)
+            .ok_or(format!("Invite '{}' not found", invite_code))?;
+        if let Some(credits) = credits {
+            invite::validate_user_invites_credits(self, user, credits, Some(*invite_credits))?;
+
+            if credits % credits_per_user != 0 {
+                return Err(format!(
+                    "Credits per user are not a multiplier of credits {} {}",
+                    credits, credits_per_user
+                ));
+            }
+        }
+
+        self.invite_codes
+            .get_mut(&invite_code)
+            .ok_or(format!("Invite '{}' not found", invite_code))?
+            .update(credits, realm_id, user_id)?;
```

Comment:
Would it make sense for `update` to run `validate_user_invites_credits` first? So that the caller doesn't need to think about validation? (this sounds more robust to me)

Reply:
validate_user_invites_credits uses state, while invite update only uses self (invite instance)

What should be done here in general? I've seen different approaches in codebase.

Reply:
Oh I see the problem. Yeah, this is an ugly artifact of Rust... No, then let it as is.